### PR TITLE
Bot permissions

### PR DESCRIPTION
**Pull request by Natshirt**
Catch error when client has missing permission
created class MissingPermissionResponse and initialize class in Command.exec() when discord.Forbidden is raised

### CODE REVIEW

File: DemonOverlord/core/util/responses.py [hunk 1]
```diff
@@ -86,3 +87,42 @@ def __init__(self, command):
...
+            "send", "kick", "ban", "unban", "edit", "delete", "purge", "webhooks",
+            "create_text_channel", "fetch_member", "bans", "prune_members", "estimate_pruned_members",
+            "invites", "create_integration", "integrations", "create_custom_emoji", "create_role",
+            "edit_role_positions", "vanity_invite", "audit_logs", "widget", "sync", "add_roles",
+            "remove_roles", "publish", "pin", "unpin"
+        )
+
+        # compile regex string to find the function who raises the discord.Forbidden
+        to_find = "in ("
+        for func in possible_functions:
```
Comment (Dragonsight91) [inline DemonOverlord/core/util/responses.py:113]:
```python
for func in possible_functions:
    to_find += func + "|"
```

could be shortened to 
```python
"|".join(func)
```

File: DemonOverlord/core/util/command.py [hunk 2]
```diff
@@ -80,9 +81,11 @@ async def exec(self) -> None:
 
             else:
                 response = BadCommandResponse(self)
+        except discord.Forbidden:
```
Comment (Dragonsight91) [inline DemonOverlord/core/util/command.py:84]: interesting way to do this... i didn't actually think about this

File: DemonOverlord/core/util/responses.py [hunk 3]
```diff
@@ -86,3 +87,42 @@ def __init__(self, command):
...
+        found_functions = re.findall(to_find, tb)
+        forbidden_function = found_functions[len(found_functions)-1] if len(found_functions) != 0 else "MISSING FUNCTION"
+
+        self.timeout = 20
+        self.add_field(name="Full Command:", value=command.full, inline=False)
+        self.add_field(
+            name="Message",
+            value=f"""Sorry, but the bot is not allowed to use `{forbidden_function}`
+            This error only occurs, when there is a missing permission.
+            Please stop trying to use it!""",
```
Comment (Dragonsight91) [inline DemonOverlord/core/util/responses.py:127]: maybe reformulate to "Please fix the permissions and add 'needed permission' to the bot role, before trying again". this way sort of discourages users from ever trying it again

File: DemonOverlord/core/util/responses.py [hunk 4]
```diff
@@ -86,3 +87,42 @@ def __init__(self, command):
...
+    This Represents a Discord Embed and any properties of that embed are active and usable by this class.
+    This class is used when the client user has a missing permission
+    """
+    def __init__(self, command, tb):
+        super().__init__(
+            f"ERROR - MISSING PERMISSION", color=0xFF0000, icon="🚫"
+        )
+
+        # discord.py functions which can raise discord.Forbidden
+        possible_functions = (
```
Comment (Dragonsight91) [inline DemonOverlord/core/util/responses.py:103]: i mean.... that's one way to do it... might be annoying to maintain, i'll have to see how this behaves in the future.

**Review by Dragonsight91 — CHANGES_REQUESTED** [review]
there is pretty much only one change that i would say is necessary before merging this change and that is already addressed in the review.

**Comment by Natshirt** [comment]
I applied the requested changes

File: DemonOverlord/core/util/responses.py [hunk 5]
```diff
@@ -99,30 +99,17 @@ def __init__(self, command, tb):
...
         self.timeout = 20
         self.add_field(name="Full Command:", value=command.full, inline=False)
         self.add_field(
             name="Message",
             value=f"""Sorry, but the bot is not allowed to use `{forbidden_function}`
             This error only occurs, when there is a missing permission.
-            Please stop trying to use it!""",
-        )
\ No newline at end of file
+            Please fix the permissions and add 'needed permission' to the bot role, before trying again.""",
```
Comment (Dragonsight91) [inline DemonOverlord/core/util/responses.py:114]: the `'necessary permissions'` was more meant as an "insert the necessary permission in that place to show the user what they gotta do."

File: DemonOverlord/core/util/responses.py [hunk 6]
```diff
@@ -99,30 +99,17 @@ def __init__(self, command, tb):
...
-        # compile regex string to find the function who raises the discord.Forbidden
-        to_find = "in ("
-        for func in possible_functions:
-            to_find += func + "|"
-        to_find = to_find[0:len(to_find)-2] + ")"
-        to_find = re.compile(r"{0}".format(to_find))
-
-        found_functions = re.findall(to_find, tb)
-        forbidden_function = found_functions[len(found_functions)-1] if len(found_functions) != 0 else "MISSING FUNCTION"
+        # use regex to find function which raise discord.Forbidden
```
Comment (Dragonsight91) [inline DemonOverlord/core/util/responses.py:102]: looks like a much more robust solution, change accepted

**Review by Dragonsight91 — CHANGES_REQUESTED** [review]
mostly approved, just the final response text was ... misunderstood

**Review by Dragonsight91 — APPROVED** [review]

**Review by Dragonsight91 — APPROVED** [review]
all changes approved

**Pull request merged**